File: InMotionGIT.Common - Copy/DataAcceess/WhereStatements/EffectiveDateCondition.vb
Namespace DataAccess.WhereStatements

    Public Class EffectiveDateCondition(Of t)
        Inherits DateCondition(Of t)

        Protected Friend ReadOnly nullfieldName As String

        Public Sub New(ByVal name As String, ByVal nulldate As String, connectionName As String, ByVal base As IWhereStatement)
            MyBase.New(name, connectionName, base)
            nullfieldName = nulldate
        End Sub

        Public Sub New(ByVal name As String, ByVal nulldate As String, ByVal base As IWhereStatement)
            MyBase.New(name, base)
            nullfieldName = nulldate
        End Sub

        Public Sub New(ByVal name As String, ByVal base As IWhereStatement)
            MyBase.New(name, base)
        End Sub

        Public Function ValidAt(value As Date) As LogicalOperator(Of t)
            tableType.command &= String.Format("{0} <= '{2}' AND ({1} IS NULL OR {1} > '{2}')", fieldName, nullfieldName, Helpers.DataAccessLayer.DateValueWithFormat(ConnectionName, value))
            Return New LogicalOperator(Of t)(tableType)
        End Function

    End Class

End Namespace

File: Common/InMotionGIT.Common/DataAcceess/WhereStatements/EffectiveDateCondition.cs
using System;

namespace InMotionGIT.Common.DataAccess.WhereStatements
{

    public class EffectiveDateCondition<t> : DateCondition<t>
    {

        protected internal readonly string nullfieldName;

        public EffectiveDateCondition(string name, string nulldate, string connectionName, IWhereStatement @base) : base(name, connectionName, @base)
        {
            nullfieldName = nulldate;
        }

        public EffectiveDateCondition(string name, string nulldate, IWhereStatement @base) : base(name, @base)
        {
            nullfieldName = nulldate;
        }

        public EffectiveDateCondition(string name, IWhereStatement @base) : base(name, @base)
        {
        }

        public LogicalOperator<t> ValidAt(DateTime value)
        {
            tableType.command += string.Format("{0} <= '{2}' AND ({1} IS NULL OR {1} > '{2}')", fieldName, nullfieldName, Helpers.DataAccessLayer.DateValueWithFormat(ConnectionName, value));
            return new LogicalOperator<t>(tableType);
        }

    }

}